src/host/inputBuffer.cpp

```diff
@@ -261,47 +402,105 @@ void InputBuffer::PassThroughWin32MouseRequest(bool enable)
         {
-            if (!WaitForData)
+            const auto keyEvent = static_cast<KeyEvent*>(event.get());
+            WORD repeat = 1;
+
+            // for stream reads we need to split any key events that have been coalesced
+            if (Stream)
             {
-                return STATUS_SUCCESS;
+                repeat = keyEvent->GetRepeatCount();
+                keyEvent->SetRepeatCount(1);
             }
-            return CONSOLE_STATUS_WAIT;
-        }
 
-        // read from buffer
-        std::deque<std::unique_ptr<IInputEvent>> events;
-        size_t eventsRead;
-        bool resetWaitEvent;
-        _ReadBuffer(events,
-                    AmountToRead,
-                    eventsRead,
-                    Peek,
-                    resetWaitEvent,
-                    Unicode,
-                    Stream);
-
-        // copy events to outEvents
-        while (!events.empty())
-        {
-            OutEvents.push_back(std::move(events.front()));
-            events.pop_front();
-        }
+            if (Unicode)
+            {
+                do
+                {
+                    OutEvents.push_back(std::make_unique<KeyEvent>(*keyEvent));
+                    repeat--;
+                } while (repeat > 0 && OutEvents.size() < AmountToRead);
+            }
+            else
+            {
+                const auto wch = keyEvent->GetCharData();
+
+                char buffer[8];
+                const auto length = WideCharToMultiByte(cp, 0, &wch, 1, &buffer[0], sizeof(buffer), nullptr, nullptr);
+                THROW_LAST_ERROR_IF(length <= 0);
+
+                const std::string_view str{ &buffer[0], gsl::narrow_cast<size_t>(length) };
+
+                do
+                {
+                    for (const auto& ch : str)
+                    {
+                        auto tempEvent = std::make_unique<KeyEvent>(*keyEvent);
+                        tempEvent->SetCharData(ch);
+                        OutEvents.push_back(std::move(tempEvent));
+                    }
+                    repeat--;
+                } while (repeat > 0 && OutEvents.size() < AmountToRead);
+            }
 
-        if (resetWaitEvent)
+            if (repeat && !Peek)
+            {
+                const auto originalKeyEvent = static_cast<KeyEvent*>((*it).get());
+                originalKeyEvent->SetRepeatCount(repeat);
+                break;
+            }
+        }
+        else
         {
-            ServiceLocator::LocateGlobals().hInputEvent.ResetEvent();
+            OutEvents.push_back(std::move(event));
         }
-        return STATUS_SUCCESS;
+
+        ++it;
     }
-    catch (...)
+
+    if (!Peek)
+    {
+        _storage.erase(_storage.begin(), it);
+    }
+
+    Cache(Unicode, OutEvents, AmountToRead);
+
+    if (OutEvents.empty())
+    {
+        return WaitForData ? CONSOLE_STATUS_WAIT : STATUS_SUCCESS;
+    }
+    if (_storage.empty())

```

Comment:
are we sure we want to reset the event when we are mptied out?

are we totally certain that the original boolean wasn't _useful_?

Reply:
It comforts me that you got confused by this code as well. Holy moly.
`resetWaitEvent` used to be an out parameter of the `_ReadBuffer` function, which had only 1 caller: this function (`Read`). `resetWaitEvent` was set like this by `_ReadBuffer`:
```cpp
resetWaitEvent = true;

// ...

// signal if we emptied the buffer
if (_storage.empty())
{
    resetWaitEvent = true;
}
```

`Read` in turn did this:
```cpp
if (resetWaitEvent)
{
    ServiceLocator::LocateGlobals().hInputEvent.ResetEvent();
}
```

By merging `_ReadBuffer` into `Read` the `resetWaitEvent` is gone and got replaced with `_storage.empty()`.